### ISSUE TITLE
Implement APFL

### ISSUE THREAD

**Issue by kdmalc** [kdmalc/personalization-privacy-risk#2]
Probably in parallel to FedAvg

**Comment by kdmalc** [comment]
They use a PyTorch variant they made for FL: https://github.com/MLOPTPSU/FedTorch

They make heavy use of loss.backward() and such to autocompute the gradient.  I don't think I can implement that myself.  Not sure what the difference between v and w (local and global) gradients are

**Comment by kdmalc** [comment]
I mostly implemented it, writing it from scratch, following their paper.  However, my L (L-smoothness) is like 40M which makes things overflow and blow up.  Not sure why this is (the norm of FF.T is just massive, but not sure why).  Is this a related issue to eta grad step which also blew up?  Normalize/clip the gradient maybe?

**Comment by kdmalc** [comment]
Normalizing F fixed L to a range of 20-300, which is better (not sure if it's correct).  But it still blows up, I'm assuming because of the gradient.  Adaptive alpha also eventually blows up.  Try setting to 0.25 like they did?

**Comment by kdmalc** [comment]
The paper: https://arxiv.org/pdf/2003.13461.pdf
Algorithm is on page 8

**Comment by kdmalc** [comment]
Superseded by PFL Non-IID codebase (has their own implementation of APFL).
